File: platform/android/app/src/main/java/pe/gob/onpe/votodigital/cifrador/android/AndroidTrueRngSupport.java
package pe.gob.onpe.votodigital.cifrador.android;

import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbManager;
import com.hoho.android.usbserial.driver.UsbSerialDriver;
import com.hoho.android.usbserial.driver.UsbSerialProber;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.logging.Logger;

public final class AndroidTrueRngSupport {

    public static final String ACTION_USB_PERMISSION =
            "pe.gob.onpe.votodigital.cifrador.android.USB_PERMISSION";
    public static final int TRUE_RNG_VENDOR_ID = 0x04d8;
    public static final int TRUE_RNG_PRODUCT_ID = 0xf5fe;

    private final Context context;
    private final UsbManager usbManager;
    private final UsbSerialProber prober;

    public AndroidTrueRngSupport(Context context) {
        this.context = context;
        this.usbManager = context.getSystemService(UsbManager.class);
        this.prober = UsbSerialProber.getDefaultProber();
    }

    public DeviceStatus scanStatus() {
        UsbDevice device = findDevice();
        if (device == null) {
            return new DeviceStatus(
                    "TrueRNG USB no detectado. Conectalo por OTG y vuelve a intentar.",
                    false,
                    false
            );
        }
        UsbSerialDriver driver = findDriver(device);
        boolean hasPermission = usbManager.hasPermission(device);
        String message = new StringBuilder()
                .append(String.format(Locale.US,
                        "TrueRNG detectado: vid=0x%04x pid=0x%04x",
                        device.getVendorId(),
                        device.getProductId()))
                .append("\ndriver=")
                .append(driver == null ? "no_serial_driver" : driver.getClass().getSimpleName())
                .append("\npermiso=")
                .append(hasPermission ? "concedido" : "pendiente")
                .toString();
        return new DeviceStatus(message, true, hasPermission);
    }

    public PermissionRequest requestPermission(PendingIntent pendingIntent) {
        UsbDevice device = findDevice();
        if (device == null) {
            return new PermissionRequest(
                    "No se encontro un TrueRNG USB conectado para pedir permiso.",
                    false
            );
        }
        if (usbManager.hasPermission(device)) {
            return new PermissionRequest(
                    "El permiso USB para TrueRNG ya estaba concedido.",
                    false
            );
        }
        usbManager.requestPermission(device, pendingIntent);
        return new PermissionRequest(
                "Solicitud de permiso USB enviada para TrueRNG.",
                true
        );
    }

    public String buildPermissionResult(UsbDevice device, boolean granted) {
        String label = device == null
                ? "TrueRNG"
                : String.format(Locale.US, "vid=0x%04x pid=0x%04x", device.getVendorId(), device.getProductId());
        return granted
                ? "Permiso USB concedido para " + label + "."
                : "Permiso USB denegado para " + label + ".";
    }

    public AndroidUsbTrueRngRandomSource openRandomSource(Logger logger) {
        UsbDevice device = findDevice();
        if (device == null) {
            throw new IllegalStateException("No se detecto un TrueRNG USB conectado en Android.");
        }
        if (!usbManager.hasPermission(device)) {
            throw new IllegalStateException(
                    "Falta permiso USB para el TrueRNG en Android. Usa 'Solicitar permiso TrueRNG'."
            );
        }
        UsbSerialDriver driver = findDriver(device);
        if (driver == null) {
            throw new IllegalStateException(
                    "Android detecto el dispositivo, pero no encontro driver serial compatible para TrueRNG."
            );
        }
        android.hardware.usb.UsbDeviceConnection connection = usbManager.openDevice(device);
        if (connection == null) {
            throw new IllegalStateException("Android no pudo abrir la conexion USB del TrueRNG.");
        }
        return AndroidUsbTrueRngRandomSource.open(device, driver, connection, logger);
    }

    public UsbDevice findDevice() {
        Map<String, UsbDevice> devices = usbManager.getDeviceList();
        for (UsbDevice device : devices.values()) {
            if (device.getVendorId() == TRUE_RNG_VENDOR_ID
                    && device.getProductId() == TRUE_RNG_PRODUCT_ID) {
                return device;
            }
        }
        return null;
    }

    private UsbSerialDriver findDriver(UsbDevice device) {
        List<UsbSerialDriver> drivers = prober.findAllDrivers(usbManager);
        for (UsbSerialDriver driver : drivers) {
            if (driver.getDevice().getDeviceId() == device.getDeviceId()) {
                return driver;
            }
        }
        return null;
    }

    public static PendingIntent buildPermissionIntent(Context context) {
        int flags = PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE;
        return PendingIntent.getBroadcast(
                context,
                2001,
                new Intent(ACTION_USB_PERMISSION),
                flags
        );
    }

    public static final class DeviceStatus {
        private final String message;
        private final boolean deviceFound;
        private final boolean permissionGranted;

        public DeviceStatus(String message, boolean deviceFound, boolean permissionGranted) {
            this.message = message;
            this.deviceFound = deviceFound;
            this.permissionGranted = permissionGranted;
        }

        public String getMessage() {
            return message;
        }

        public boolean getDeviceFound() {
            return deviceFound;
        }

        public boolean getPermissionGranted() {
            return permissionGranted;
        }
    }

    public static final class PermissionRequest {
        private final String message;
        private final boolean requested;

        public PermissionRequest(String message, boolean requested) {
            this.message = message;
            this.requested = requested;
        }

        public String getMessage() {
            return message;
        }

        public boolean getRequested() {
            return requested;
        }
    }
}

File: android/app/src/main/java/pe/gob/onpe/votodigital/cifrador/android/AndroidTrueRngSupport.kt
package pe.gob.onpe.votodigital.cifrador.android

import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.hardware.usb.UsbDevice
import android.hardware.usb.UsbManager
import com.hoho.android.usbserial.driver.UsbSerialDriver
import com.hoho.android.usbserial.driver.UsbSerialProber

class AndroidTrueRngSupport(private val context: Context) {

    data class DeviceStatus(
        val message: String,
        val deviceFound: Boolean,
        val permissionGranted: Boolean
    )

    data class PermissionRequest(
        val message: String,
        val requested: Boolean
    )

    private val usbManager = context.getSystemService(UsbManager::class.java)
    private val prober = UsbSerialProber.getDefaultProber()

    fun scanStatus(): DeviceStatus {
        val device = findDevice()
            ?: return DeviceStatus(
                message = "TrueRNG USB no detectado. Conectalo por OTG y vuelve a intentar.",
                deviceFound = false,
                permissionGranted = false
            )
        val driver = findDriver(device)
        val hasPermission = usbManager.hasPermission(device)
        val message = buildString {
            append("TrueRNG detectado: vid=0x%04x pid=0x%04x".format(device.vendorId, device.productId))
            append("\ndriver=")
            append(driver?.javaClass?.simpleName ?: "no_serial_driver")
            append("\npermiso=")
            append(if (hasPermission) "concedido" else "pendiente")
        }
        return DeviceStatus(message, true, hasPermission)
    }

    fun requestPermission(pendingIntent: PendingIntent): PermissionRequest {
        val device = findDevice()
            ?: return PermissionRequest(
                message = "No se encontro un TrueRNG USB conectado para pedir permiso.",
                requested = false
            )
        if (usbManager.hasPermission(device)) {
            return PermissionRequest(
                message = "El permiso USB para TrueRNG ya estaba concedido.",
                requested = false
            )
        }
        usbManager.requestPermission(device, pendingIntent)
        return PermissionRequest(
            message = "Solicitud de permiso USB enviada para TrueRNG.",
            requested = true
        )
    }

    fun buildPermissionResult(device: UsbDevice?, granted: Boolean): String {
        val label = if (device == null) {
            "TrueRNG"
        } else {
            "vid=0x%04x pid=0x%04x".format(device.vendorId, device.productId)
        }
        return if (granted) {
            "Permiso USB concedido para $label."
        } else {
            "Permiso USB denegado para $label."
        }
    }

    fun openRandomSource(logger: java.util.logging.Logger): AndroidUsbTrueRngRandomSource {
        val device = findDevice()
            ?: throw IllegalStateException("No se detecto un TrueRNG USB conectado en Android.")
        require(usbManager.hasPermission(device)) {
            "Falta permiso USB para el TrueRNG en Android. Usa 'Solicitar permiso TrueRNG'."
        }
        val driver = findDriver(device)
            ?: throw IllegalStateException("Android detecto el dispositivo, pero no encontro driver serial compatible para TrueRNG.")
        val connection = usbManager.openDevice(device)
            ?: throw IllegalStateException("Android no pudo abrir la conexion USB del TrueRNG.")
        return AndroidUsbTrueRngRandomSource.open(device, driver, connection, logger)
    }

    fun findDevice(): UsbDevice? {
        return usbManager.deviceList.values.firstOrNull {
            it.vendorId == TRUE_RNG_VENDOR_ID && it.productId == TRUE_RNG_PRODUCT_ID
        }
    }

    private fun findDriver(device: UsbDevice): UsbSerialDriver? {
        return prober.findAllDrivers(usbManager).firstOrNull { it.device.deviceId == device.deviceId }
    }

    companion object {
        const val ACTION_USB_PERMISSION = "pe.gob.onpe.votodigital.cifrador.android.USB_PERMISSION"
        const val TRUE_RNG_VENDOR_ID = 0x04d8
        const val TRUE_RNG_PRODUCT_ID = 0xf5fe

        fun buildPermissionIntent(context: Context): PendingIntent {
            val flags = PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            return PendingIntent.getBroadcast(
                context,
                2001,
                Intent(ACTION_USB_PERMISSION),
                flags
            )
        }
    }
}
